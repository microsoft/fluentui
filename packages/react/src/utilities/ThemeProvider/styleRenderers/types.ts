import { IFontFace, IKeyframes } from '@fluentui/merge-styles';

/* eslint-disable @typescript-eslint/naming-convention */

type StyleRendererOptions = {
  rtl?: boolean;
  targetWindow: Window | undefined;
};

export interface StyleRenderer {
  /**
   * Expected to initialize or re-initialize the renderer. Primarily for testing purposes.
   */
  reset: () => void;

  /**
   * Returns a unique id for the renderer; used as part of the cache key when determining if new
   * styles need to be rendered.
   */
  getId: () => number;

  /**
   * Renders a stylesheet and returns the map of key to class name.
   */
  renderStyles: <TRuleSet>(ruleSet: TRuleSet, options: StyleRendererOptions) => { [key in keyof TRuleSet]: string };

  /**
   * Renders keyframes and returns the keyframe name.
   */
  renderKeyframes: (keyframes: IKeyframes, options: StyleRendererOptions) => string;

  /**
   * Renders fontfaces.
   */
  renderFontFace: (fontFace: IFontFace, options: StyleRendererOptions) => void;
}
