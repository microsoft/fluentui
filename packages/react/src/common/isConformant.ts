import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant<TProps = {}>(
  testInfo: Omit<IsConformantOptions<TProps>, 'componentPath'> & { componentPath?: string },
) {
  const defaultOptions: Partial<IsConformantOptions<TProps>> = {
    disabledTests: [
      'has-docblock',
      'kebab-aria-attributes',
      'is-static-property-of-parent',
      // Will enable with appropriate overrides separately
      'consistent-callback-names',
    ],
    skipAsPropTests: true,
    componentPath: module!.parent!.filename.replace('.test', ''),
  };

  baseIsConformant(defaultOptions, testInfo);
}
