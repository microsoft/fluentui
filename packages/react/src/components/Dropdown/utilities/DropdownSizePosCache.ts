import { IDropdownOption, DropdownMenuItemType } from '../Dropdown.types';

/**
 * A utility class to cache size and position in cache.
 *
 * Dropdown options has non-selectable display types. It is therefore not cheap to determine
 * the total number of actual selectable options as well as the position an option is in the
 * list of options - O(n) cost for each lookup.
 *
 * Given that we potentially have to make this determination on every single render pass, this
 * cache should provide a little bit of relief.
 */
export class DropdownSizePosCache {
  private _cachedOptions: IDropdownOption[];
  private _displayOnlyOptionsCache: number[];
  private _size = 0;

  /**
   * Invalidates the cache and recalculate the size of selectable options.
   */
  public updateOptions(options: IDropdownOption[]) {
    const displayOnlyOptionsCache = [];
    let size = 0;
    for (let i = 0; i < options.length; i++) {
      if (options[i].itemType === DropdownMenuItemType.Divider || options[i].itemType === DropdownMenuItemType.Header) {
        displayOnlyOptionsCache.push(i);
      } else if (!options[i].hidden) {
        size++;
      }
    }

    this._size = size;
    this._displayOnlyOptionsCache = displayOnlyOptionsCache;
    this._cachedOptions = [...options];
  }

  /**
   * The size of all the selectable options.
   */
  public get optionSetSize(): number {
    return this._size;
  }

  /**
   * The chached options array.
   */
  public get cachedOptions(): IDropdownOption[] {
    return this._cachedOptions;
  }

  /**
   * Returns the position of this option element relative to the full set of selectable option elements.
   * Note: the first selectable element is position 1 in the set.
   * @param index The raw index of the option element.
   */
  public positionInSet(index: number | undefined): number | undefined {
    if (index === undefined) {
      return undefined;
    }

    // we could possibly memoize this too but this should be good enough, most of the time (the expectation is that
    // when you have a lot of options, the selectable options will heavily dominate over the non-selectable options.
    let offset = 0;
    while (index > this._displayOnlyOptionsCache[offset]) {
      offset++;
    }

    if (this._displayOnlyOptionsCache[offset] === index) {
      throw new Error(`Unexpected: Option at index ${index} is not a selectable element.`);
    }

    return index - offset + 1;
  }
}
