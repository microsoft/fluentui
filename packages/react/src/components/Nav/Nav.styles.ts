import { INavStyleProps, INavStyles } from './Nav.types';
import { IButtonStyles } from '../../Button';
import {
  AnimationClassNames,
  getFocusStyle,
  ZIndexes,
  getGlobalClassNames,
  HighContrastSelector,
  FontWeights,
} from '../../Styling';

const GlobalClassNames = {
  root: 'ms-Nav',
  linkText: 'ms-Nav-linkText',
  compositeLink: 'ms-Nav-compositeLink',
  link: 'ms-Nav-link',
  chevronButton: 'ms-Nav-chevronButton',
  chevronIcon: 'ms-Nav-chevron',
  navItem: 'ms-Nav-navItem',
  navItems: 'ms-Nav-navItems',
  group: 'ms-Nav-group',
  groupContent: 'ms-Nav-groupContent',
};

export const buttonStyles: IButtonStyles = {
  textContainer: {
    overflow: 'hidden',
  },
  label: {
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
  },
};

export const getStyles = (props: INavStyleProps): INavStyles => {
  const {
    className,
    theme,
    isOnTop,
    isExpanded,
    isGroup,
    isLink,
    isSelected,
    isDisabled,
    isButtonEntry,
    navHeight = 44,
    position,
    leftPadding = 20,
    leftPaddingExpanded = 28,
    rightPadding = 20,
  } = props;

  const { palette, semanticColors, fonts } = theme;

  const classNames = getGlobalClassNames(GlobalClassNames, theme);

  return {
    root: [
      classNames.root,
      className,
      fonts.medium,
      {
        overflowY: 'auto',
        userSelect: 'none',
        WebkitOverflowScrolling: 'touch',
      },
      isOnTop && [
        {
          position: 'absolute',
        },
        AnimationClassNames.slideRightIn40,
      ],
    ],
    linkText: [
      classNames.linkText,
      {
        margin: '0 4px',
        overflow: 'hidden',
        verticalAlign: 'middle',
        textAlign: 'left',
        textOverflow: 'ellipsis',
      },
    ],
    compositeLink: [
      classNames.compositeLink,
      {
        display: 'block',
        position: 'relative',
        color: semanticColors.bodyText,
      },
      isExpanded && 'is-expanded',
      isSelected && 'is-selected',
      isDisabled && 'is-disabled',
      isDisabled && {
        color: semanticColors.disabledText,
      },
    ],
    link: [
      classNames.link,
      getFocusStyle(theme),
      {
        display: 'block',
        position: 'relative',
        height: navHeight,
        width: '100%',
        lineHeight: `${navHeight}px`,
        textDecoration: 'none',
        cursor: 'pointer',
        textOverflow: 'ellipsis',
        whiteSpace: 'nowrap',
        overflow: 'hidden',
        paddingLeft: leftPadding,
        paddingRight: rightPadding,
        color: semanticColors.bodyText,
        selectors: {
          [HighContrastSelector]: {
            border: 0,
            selectors: {
              ':focus': {
                border: '1px solid WindowText',
              },
            },
          },
        },
      },
      !isDisabled && {
        selectors: {
          '.ms-Nav-compositeLink:hover &': {
            backgroundColor: semanticColors.bodyBackgroundHovered,
          },
        },
      },
      isSelected && {
        color: semanticColors.bodyTextChecked,
        fontWeight: FontWeights.semibold, // todo: get from theme
        backgroundColor: semanticColors.bodyBackgroundChecked,
        selectors: {
          '&:after': {
            borderLeft: `2px solid ${palette.themePrimary}`,
            content: '""',
            position: 'absolute',
            top: 0,
            right: 0,
            bottom: 0,
            left: 0,
            pointerEvents: 'none',
          },
        },
      },
      isDisabled && {
        color: semanticColors.disabledText,
      },
      isButtonEntry && {
        color: palette.themePrimary,
      },
    ],
    chevronButton: [
      classNames.chevronButton,
      getFocusStyle(theme),
      fonts.small,
      {
        display: 'block',
        textAlign: 'left',
        lineHeight: `${navHeight}px`,
        margin: '5px 0',
        padding: `0px, ${rightPadding}px, 0px, ${leftPaddingExpanded}px`,
        border: 'none',
        textOverflow: 'ellipsis',
        whiteSpace: 'nowrap',
        overflow: 'hidden',
        cursor: 'pointer',
        color: semanticColors.bodyText,
        backgroundColor: 'transparent',
        selectors: {
          '&:visited': {
            color: semanticColors.bodyText,
          },
        },
      },
      isGroup && {
        fontSize: fonts.large.fontSize,
        width: '100%',
        height: navHeight,
        borderBottom: `1px solid ${semanticColors.bodyDivider}`,
      },
      isLink && {
        display: 'block',
        width: leftPaddingExpanded - 2,
        height: navHeight - 2,
        position: 'absolute',
        top: '1px',
        left: `${position}px`,
        zIndex: ZIndexes.Nav,
        padding: 0,
        margin: 0,
      },
      isSelected && {
        color: palette.themePrimary,
        backgroundColor: palette.neutralLighterAlt,
        selectors: {
          '&:after': {
            borderLeft: `2px solid ${palette.themePrimary}`,
            content: '""',
            position: 'absolute',
            top: 0,
            right: 0,
            bottom: 0,
            left: 0,
            pointerEvents: 'none',
          },
        },
      },
    ],
    chevronIcon: [
      classNames.chevronIcon,
      {
        position: 'absolute',
        left: '8px',
        height: navHeight,
        // inline-flex prevents the chevron from shifting with custom line height styles
        display: 'inline-flex',
        alignItems: 'center',
        lineHeight: `${navHeight}px`,
        fontSize: fonts.small.fontSize,
        transition: 'transform .1s linear',
      },
      isExpanded && {
        transform: 'rotate(-180deg)',
      },
      isLink && {
        top: 0,
      },
    ],
    navItem: [
      classNames.navItem,
      {
        padding: 0,
      },
    ],
    navItems: [
      classNames.navItems,
      {
        listStyleType: 'none',
        padding: 0,
        margin: 0, // remove default <UL> styles
      },
    ],
    group: [classNames.group, isExpanded && 'is-expanded'],
    groupContent: [
      classNames.groupContent,
      {
        display: 'none',
        marginBottom: '40px',
      },
      AnimationClassNames.slideDownIn20,
      isExpanded && {
        display: 'block',
      },
    ],
  };
};
