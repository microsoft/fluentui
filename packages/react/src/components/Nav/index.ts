export * from './Nav';
export * from './Nav.base';
export * from './Nav.types';
