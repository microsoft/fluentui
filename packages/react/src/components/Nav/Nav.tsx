import * as React from 'react';
import { styled } from '../../Utilities';
import { INavProps, INavStyleProps, INavStyles } from './Nav.types';
import { NavBase } from './Nav.base';
import { getStyles } from './Nav.styles';

export const Nav: React.FunctionComponent<INavProps> = styled<INavProps, INavStyleProps, INavStyles>(
  NavBase,
  getStyles,
  undefined,
  {
    scope: 'Nav',
  },
);
