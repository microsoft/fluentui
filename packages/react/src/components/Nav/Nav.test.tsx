import * as React from 'react';

import { mount } from 'enzyme';
import * as renderer from 'react-test-renderer';

import { Nav } from './Nav';
import { NavBase } from './Nav.base';
import { INavLink, IRenderGroupHeaderProps, INavLinkGroup, INavButtonProps } from './Nav.types';
import { IRenderFunction, IComponentAsProps } from '@fluentui/utilities';
import { isConformant } from '../../common/isConformant';

const linkOne: INavLink = {
  key: 'Bing',
  name: 'Bing',
  url: 'http://localhost/#/testing1',
};

const linkTwo: INavLink = {
  key: 'OneDrive',
  name: 'OneDrive',
  url: 'http://localhost/#/testing2',
};

describe('Nav', () => {
  it('renders Nav correctly', () => {
    const component = renderer.create(
      <Nav
        groups={[
          {
            links: [
              {
                name: '',
                url: '',
              },
            ],
          },
        ]}
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  isConformant({
    Component: Nav,
    displayName: 'Nav',
    // Problem: Ref is not supported
    // Solution: Convert to FunctionComponent and support using forwardRef
    disabledTests: ['component-has-root-ref', 'component-handles-ref', 'component-handles-classname'],
  });

  it('render Nav with overrides correctly', () => {
    const LinkAs = (props: IComponentAsProps<INavButtonProps>): JSX.Element | null => {
      const { defaultRender: DefaultRender, ...buttonProps } = props;

      if (!DefaultRender) {
        return null;
      }

      return (
        <div data-test="button-override">
          <DefaultRender {...buttonProps} />
        </div>
      );
    };

    function onRenderNavLink(props?: INavLink, defaultRender?: IRenderFunction<INavLink>): JSX.Element | null {
      if (!props || !defaultRender) {
        return null;
      }

      return <div data-test="link-override">{defaultRender(props)}</div>;
    }

    function onRenderGroupHeader(
      props?: IRenderGroupHeaderProps,
      defaultRender?: IRenderFunction<IRenderGroupHeaderProps>,
    ): JSX.Element | null {
      if (!props || !defaultRender) {
        return null;
      }

      return <div data-test="header-override">{defaultRender(props)}</div>;
    }
    const groups: INavLinkGroup[] = [
      {
        name: 'Group',
        links: [linkOne, linkTwo],
      },
    ];

    const component = renderer.create(
      <Nav groups={groups} onRenderGroupHeader={onRenderGroupHeader} onRenderLink={onRenderNavLink} linkAs={LinkAs} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('calls onClick() correctly', () => {
    const handler = jest.fn();
    const nav = mount<NavBase>(
      <NavBase
        groups={[
          {
            links: [
              {
                name: 'foo',
                url: 'http://example.com',
                onClick: handler,
              },
            ],
          },
        ]}
      />,
    );

    const link = nav.find('a.ms-Button');
    link.simulate('click');
    expect(handler.mock.calls.length).toBe(1);
  });

  it('do not call onClick() on disabled link', () => {
    const handler = jest.fn();
    const nav = mount<NavBase>(
      <NavBase
        groups={[
          {
            links: [
              {
                name: 'foo',
                url: 'http://example.com',
                onClick: handler,
                disabled: true,
              },
            ],
          },
        ]}
      />,
    );

    const link = nav.find('button.ms-Button');
    link.simulate('click');
    expect(handler.mock.calls.length).toBe(0);
  });

  it('sets ARIA label on the nav element', () => {
    const label = 'The navigation label';
    const nav = mount<NavBase>(<Nav ariaLabel={label} groups={[]} />);
    expect(nav.find('[role="navigation"]').prop('aria-label')).toEqual(label);
  });

  it('uses location.href to determine link selected status if state/props is not set', () => {
    const props = { groups: [{ links: [linkOne, linkTwo] }] };
    const nav = mount<NavBase>(<Nav {...props} />);
    window.history.pushState({}, '', '/#/testing1');
    nav.setProps(props);

    expect(nav.getDOMNode().querySelectorAll('.ms-Nav-compositeLink.is-selected').length).toBe(1);
    expect(nav.getDOMNode().querySelectorAll('.ms-Nav-compositeLink.is-selected')[0].textContent).toEqual(linkOne.name);
  });

  it('prioritizes state over location.href to determine link selected status', () => {
    const props = { groups: [{ links: [linkOne, linkTwo] }] };
    const nav = mount<NavBase>(<Nav {...props} />);
    window.history.pushState({}, '', '/#/testing2');
    nav.setProps(props);

    nav.find('.ms-Button').first().simulate('click');
    expect(nav.getDOMNode().querySelectorAll('.ms-Nav-compositeLink.is-selected').length).toBe(1);
    expect(nav.getDOMNode().querySelectorAll('.ms-Nav-compositeLink.is-selected')[0].textContent).toEqual(linkOne.name);
  });

  it('places the correct values on rel depending on the url and target specified', () => {
    const linkWithNoTargetSpecified: INavLink = {
      key: 'Link1',
      name: 'Link1',
      url: 'https://cdpn.io/bar',
    };
    const linkWithRelativeURL: INavLink = {
      key: 'Link2',
      name: 'Link2',
      target: '_blank',
      url: '/foo',
    };
    const linkWithNonRelativeURL: INavLink = {
      key: 'Link3',
      name: 'Link3',
      target: '_blank',
      url: 'https://cdpn.io/bar',
    };
    const linkWithNonRelativeURL2: INavLink = {
      key: 'Link4',
      name: 'Link4',
      target: '_blank',
      url: 'http://cdpn.io/bar',
    };
    const nav = mount<NavBase>(
      <Nav
        groups={[
          { links: [linkWithNoTargetSpecified, linkWithRelativeURL, linkWithNonRelativeURL, linkWithNonRelativeURL2] },
        ]}
      />,
    );

    const links = nav.getDOMNode().querySelectorAll('.ms-Nav-link');
    expect(links.length).toBe(4);
    expect(links[0].getAttribute('rel')).toBeFalsy();
    expect(links[1].getAttribute('rel')).toBeFalsy();
    expect(links[2].getAttribute('rel')).toBe('noopener noreferrer');
    expect(links[3].getAttribute('rel')).toBe('noopener noreferrer');
  });
});
