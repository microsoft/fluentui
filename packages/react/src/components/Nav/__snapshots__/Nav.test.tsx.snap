// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav render Nav with overrides correctly 1`] = `
<div
  className=
      ms-FocusZone
      &:focus {
        outline: none;
      }
  data-focuszone-id="FocusZone4"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDownCapture={[Function]}
>
  <nav
    className=
        ms-Nav
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          -webkit-overflow-scrolling: touch;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          overflow-y: auto;
          user-select: none;
        }
    role="navigation"
  >
    <div
      className="ms-Nav-group is-expanded"
    >
      <div
        data-test="header-override"
      >
        <button
          aria-expanded={true}
          className=
              ms-Nav-chevronButton
              {
                -moz-osx-font-smoothing: grayscale;
                -webkit-font-smoothing: antialiased;
                background-color: transparent;
                border-bottom: 1px solid #edebe9;
                border: none;
                color: #323130;
                cursor: pointer;
                display: block;
                font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                font-size: 18px;
                font-weight: 400;
                height: 44px;
                line-height: 44px;
                margin-bottom: 5px;
                margin-left: 0;
                margin-right: 0;
                margin-top: 5px;
                outline: transparent;
                overflow: hidden;
                padding-bottom: 0px,;
                padding-left: 28px;
                padding-right: 20px,;
                padding-top: 0px,;
                position: relative;
                text-align: left;
                text-overflow: ellipsis;
                white-space: nowrap;
                width: 100%;
              }
              &::-moz-focus-inner {
                border: 0;
              }
              .ms-Fabric--isFocusVisible &:focus:after {
                border: 1px solid #ffffff;
                bottom: 1px;
                content: "";
                left: 1px;
                outline: 1px solid #605e5c;
                position: absolute;
                right: 1px;
                top: 1px;
                z-index: 1;
              }
              &:visited {
                color: #323130;
              }
          onClick={[Function]}
        >
          <i
            aria-hidden={true}
            className=
                ms-Nav-chevron
                {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  align-items: center;
                  display: inline-flex;
                  font-family: "FabricMDL2Icons";
                  font-size: 12px;
                  font-style: normal;
                  font-weight: normal;
                  height: 44px;
                  left: 8px;
                  line-height: 44px;
                  position: absolute;
                  speak: none;
                  transform: rotate(-180deg);
                  transition: transform .1s linear;
                }
            data-icon-name="ChevronDown"
          >
            Óúç
          </i>
          Group
        </button>
      </div>
      <div
        className=
            ms-Nav-groupContent
            {
              animation-duration: 0.367s;
              animation-fill-mode: both;
              animation-name: keyframes from{opacity:0;}to{opacity:1;} keyframes from{transform:translate3d(0,-20px,0);pointer-events:none;}to{transform:translate3d(0,0,0);pointer-events:auto;};
              animation-timing-function: cubic-bezier(.1,.9,.2,1);
              display: block;
              margin-bottom: 40px;
            }
      >
        <ul
          className=
              ms-Nav-navItems
              {
                list-style-type: none;
                margin-bottom: 0px;
                margin-left: 0px;
                margin-right: 0px;
                margin-top: 0px;
                padding-bottom: 0px;
                padding-left: 0px;
                padding-right: 0px;
                padding-top: 0px;
              }
          role="list"
        >
          <li
            className=
                ms-Nav-navItem
                {
                  padding-bottom: 0px;
                  padding-left: 0px;
                  padding-right: 0px;
                  padding-top: 0px;
                }
            role="listitem"
          >
            <div
              className=
                  ms-Nav-compositeLink
                  {
                    color: #323130;
                    display: block;
                    position: relative;
                  }
              name="Bing"
            >
              <div
                data-test="button-override"
              >
                <a
                  className=
                      ms-Button
                      ms-Button--action
                      ms-Button--command
                      ms-Nav-link
                      {
                        -moz-osx-font-smoothing: grayscale;
                        -webkit-font-smoothing: antialiased;
                        background-color: transparent;
                        border-radius: 2px;
                        border: 1px solid transparent;
                        box-sizing: border-box;
                        color: #323130;
                        cursor: pointer;
                        display: block;
                        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                        font-size: 14px;
                        font-weight: 400;
                        height: 44px;
                        line-height: 44px;
                        outline: transparent;
                        overflow: hidden;
                        padding-bottom: 0;
                        padding-left: 27px;
                        padding-right: 20px;
                        padding-top: 0;
                        position: relative;
                        text-align: center;
                        text-decoration: none;
                        text-overflow: ellipsis;
                        user-select: none;
                        white-space: nowrap;
                        width: 100%;
                      }
                      &::-moz-focus-inner {
                        border: 0;
                      }
                      .ms-Fabric--isFocusVisible &:focus:after {
                        border: 1px solid #ffffff;
                        bottom: 1px;
                        content: "";
                        left: 1px;
                        outline: 1px solid #605e5c;
                        position: absolute;
                        right: 1px;
                        top: 1px;
                        z-index: 1;
                      }
                      @media screen and (-ms-high-contrast: active), (forced-colors: active){.ms-Fabric--isFocusVisible &:focus:after {
                        bottom: -2px;
                        left: -2px;
                        outline-color: ButtonText;
                        right: -2px;
                        top: -2px;
                      }
                      &:active > * {
                        left: 0px;
                        position: relative;
                        top: 0px;
                      }
                      @media screen and (-ms-high-contrast: active), (forced-colors: active){& {
                        border-color: Window;
                        border: 0px;
                      }
                      &:hover {
                        color: #0078d4;
                      }
                      @media screen and (-ms-high-contrast: active), (forced-colors: active){&:hover {
                        color: Highlight;
                      }
                      &:hover .ms-Button-icon {
                        color: #0078d4;
                      }
                      &:active {
                        color: #000000;
                      }
                      &:active .ms-Button-icon {
                        color: #004578;
                      }
                      @media screen and (-ms-high-contrast: active), (forced-colors: active){&:focus {
                        border: 1px solid WindowText;
                      }
                      .ms-Nav-compositeLink:hover & {
                        background-color: #f3f2f1;
                      }
                  data-is-focusable={true}
                  href="http://localhost/#/testing1"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  onKeyPress={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                  title="Bing"
                >
                  <span
                    className=
                        ms-Button-flexContainer
                        {
                          align-items: center;
                          display: flex;
                          flex-wrap: nowrap;
                          height: 100%;
                          justify-content: flex-start;
                        }
                    data-automationid="splitbuttonprimary"
                  >
                    <div
                      data-test="link-override"
                    >
                      <div
                        className=
                            ms-Nav-linkText
                            {
                              margin-bottom: 0;
                              margin-left: 4px;
                              margin-right: 4px;
                              margin-top: 0;
                              overflow: hidden;
                              text-align: left;
                              text-overflow: ellipsis;
                              vertical-align: middle;
                            }
                      >
                        Bing
                      </div>
                    </div>
                  </span>
                </a>
              </div>
            </div>
          </li>
          <li
            className=
                ms-Nav-navItem
                {
                  padding-bottom: 0px;
                  padding-left: 0px;
                  padding-right: 0px;
                  padding-top: 0px;
                }
            role="listitem"
          >
            <div
              className=
                  ms-Nav-compositeLink
                  {
                    color: #323130;
                    display: block;
                    position: relative;
                  }
              name="OneDrive"
            >
              <div
                data-test="button-override"
              >
                <a
                  className=
                      ms-Button
                      ms-Button--action
                      ms-Button--command
                      ms-Nav-link
                      {
                        -moz-osx-font-smoothing: grayscale;
                        -webkit-font-smoothing: antialiased;
                        background-color: transparent;
                        border-radius: 2px;
                        border: 1px solid transparent;
                        box-sizing: border-box;
                        color: #323130;
                        cursor: pointer;
                        display: block;
                        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                        font-size: 14px;
                        font-weight: 400;
                        height: 44px;
                        line-height: 44px;
                        outline: transparent;
                        overflow: hidden;
                        padding-bottom: 0;
                        padding-left: 27px;
                        padding-right: 20px;
                        padding-top: 0;
                        position: relative;
                        text-align: center;
                        text-decoration: none;
                        text-overflow: ellipsis;
                        user-select: none;
                        white-space: nowrap;
                        width: 100%;
                      }
                      &::-moz-focus-inner {
                        border: 0;
                      }
                      .ms-Fabric--isFocusVisible &:focus:after {
                        border: 1px solid #ffffff;
                        bottom: 1px;
                        content: "";
                        left: 1px;
                        outline: 1px solid #605e5c;
                        position: absolute;
                        right: 1px;
                        top: 1px;
                        z-index: 1;
                      }
                      @media screen and (-ms-high-contrast: active), (forced-colors: active){.ms-Fabric--isFocusVisible &:focus:after {
                        bottom: -2px;
                        left: -2px;
                        outline-color: ButtonText;
                        right: -2px;
                        top: -2px;
                      }
                      &:active > * {
                        left: 0px;
                        position: relative;
                        top: 0px;
                      }
                      @media screen and (-ms-high-contrast: active), (forced-colors: active){& {
                        border-color: Window;
                        border: 0px;
                      }
                      &:hover {
                        color: #0078d4;
                      }
                      @media screen and (-ms-high-contrast: active), (forced-colors: active){&:hover {
                        color: Highlight;
                      }
                      &:hover .ms-Button-icon {
                        color: #0078d4;
                      }
                      &:active {
                        color: #000000;
                      }
                      &:active .ms-Button-icon {
                        color: #004578;
                      }
                      @media screen and (-ms-high-contrast: active), (forced-colors: active){&:focus {
                        border: 1px solid WindowText;
                      }
                      .ms-Nav-compositeLink:hover & {
                        background-color: #f3f2f1;
                      }
                  data-is-focusable={true}
                  href="http://localhost/#/testing2"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  onKeyPress={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                  title="OneDrive"
                >
                  <span
                    className=
                        ms-Button-flexContainer
                        {
                          align-items: center;
                          display: flex;
                          flex-wrap: nowrap;
                          height: 100%;
                          justify-content: flex-start;
                        }
                    data-automationid="splitbuttonprimary"
                  >
                    <div
                      data-test="link-override"
                    >
                      <div
                        className=
                            ms-Nav-linkText
                            {
                              margin-bottom: 0;
                              margin-left: 4px;
                              margin-right: 4px;
                              margin-top: 0;
                              overflow: hidden;
                              text-align: left;
                              text-overflow: ellipsis;
                              vertical-align: middle;
                            }
                      >
                        OneDrive
                      </div>
                    </div>
                  </span>
                </a>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</div>
`;

exports[`Nav renders Nav correctly 1`] = `
<div
  className=
      ms-FocusZone
      &:focus {
        outline: none;
      }
  data-focuszone-id="FocusZone0"
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDownCapture={[Function]}
>
  <nav
    className=
        ms-Nav
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          -webkit-overflow-scrolling: touch;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          overflow-y: auto;
          user-select: none;
        }
    role="navigation"
  >
    <div
      className="ms-Nav-group is-expanded"
    >
      <div
        className=
            ms-Nav-groupContent
            {
              animation-duration: 0.367s;
              animation-fill-mode: both;
              animation-name: keyframes from{opacity:0;}to{opacity:1;} keyframes from{transform:translate3d(0,-20px,0);pointer-events:none;}to{transform:translate3d(0,0,0);pointer-events:auto;};
              animation-timing-function: cubic-bezier(.1,.9,.2,1);
              display: block;
              margin-bottom: 40px;
            }
      >
        <ul
          className=
              ms-Nav-navItems
              {
                list-style-type: none;
                margin-bottom: 0px;
                margin-left: 0px;
                margin-right: 0px;
                margin-top: 0px;
                padding-bottom: 0px;
                padding-left: 0px;
                padding-right: 0px;
                padding-top: 0px;
              }
          role="list"
        >
          <li
            className=
                ms-Nav-navItem
                {
                  padding-bottom: 0px;
                  padding-left: 0px;
                  padding-right: 0px;
                  padding-top: 0px;
                }
            role="listitem"
          >
            <div
              className=
                  ms-Nav-compositeLink
                  {
                    color: #323130;
                    display: block;
                    position: relative;
                  }
              name=""
            >
              <button
                className=
                    ms-Button
                    ms-Button--action
                    ms-Button--command
                    ms-Nav-link
                    {
                      -moz-osx-font-smoothing: grayscale;
                      -webkit-font-smoothing: antialiased;
                      background-color: transparent;
                      border-radius: 2px;
                      border: 1px solid transparent;
                      box-sizing: border-box;
                      color: #323130;
                      cursor: pointer;
                      display: block;
                      font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                      font-size: 14px;
                      font-weight: 400;
                      height: 44px;
                      line-height: 44px;
                      outline: transparent;
                      overflow: hidden;
                      padding-bottom: 0;
                      padding-left: 27px;
                      padding-right: 20px;
                      padding-top: 0;
                      position: relative;
                      text-align: center;
                      text-decoration: none;
                      text-overflow: ellipsis;
                      user-select: none;
                      white-space: nowrap;
                      width: 100%;
                    }
                    &::-moz-focus-inner {
                      border: 0;
                    }
                    .ms-Fabric--isFocusVisible &:focus:after {
                      border: 1px solid #ffffff;
                      bottom: 1px;
                      content: "";
                      left: 1px;
                      outline: 1px solid #605e5c;
                      position: absolute;
                      right: 1px;
                      top: 1px;
                      z-index: 1;
                    }
                    @media screen and (-ms-high-contrast: active), (forced-colors: active){.ms-Fabric--isFocusVisible &:focus:after {
                      bottom: -2px;
                      left: -2px;
                      outline-color: ButtonText;
                      right: -2px;
                      top: -2px;
                    }
                    &:active > * {
                      left: 0px;
                      position: relative;
                      top: 0px;
                    }
                    @media screen and (-ms-high-contrast: active), (forced-colors: active){& {
                      border-color: Window;
                      border: 0px;
                    }
                    &:hover {
                      color: #0078d4;
                    }
                    @media screen and (-ms-high-contrast: active), (forced-colors: active){&:hover {
                      color: Highlight;
                    }
                    &:hover .ms-Button-icon {
                      color: #0078d4;
                    }
                    &:active {
                      color: #000000;
                    }
                    &:active .ms-Button-icon {
                      color: #004578;
                    }
                    @media screen and (-ms-high-contrast: active), (forced-colors: active){&:focus {
                      border: 1px solid WindowText;
                    }
                    .ms-Nav-compositeLink:hover & {
                      background-color: #f3f2f1;
                    }
                data-is-focusable={true}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                title=""
                type="button"
              >
                <span
                  className=
                      ms-Button-flexContainer
                      {
                        align-items: center;
                        display: flex;
                        flex-wrap: nowrap;
                        height: 100%;
                        justify-content: flex-start;
                      }
                  data-automationid="splitbuttonprimary"
                >
                  <div
                    className=
                        ms-Nav-linkText
                        {
                          margin-bottom: 0;
                          margin-left: 4px;
                          margin-right: 4px;
                          margin-top: 0;
                          overflow: hidden;
                          text-align: left;
                          text-overflow: ellipsis;
                          vertical-align: middle;
                        }
                  >
                    
                  </div>
                </span>
              </button>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</div>
`;
